### PR TITLE
Add location and sign to KS test result

### LINKED ISSUE #9389: Kolmogorov Smirnov 2 samples returning max distance location [Enhancement]
2 sample Kolmogorov Smirnov test (KS test) is about finding the maximal vertical distance between two empirical cumulative distributions; a large max distance is unlikely under the null assumption that the two empirical distributions are sampled from the same theoretical curve. This is implemented in `scipy.stats.ks_twosamp`.

The function calculates but does not return the location of the cumulative at which the max distance happens, but that would be extremely useful for a number of machine learning algorithms, e.g. to get an initial guess for a classifier between the two kinds of data.

I propose to add an optional argument to the function to return the location of the max distance between the distributions in terms of indices of the data and also value of the quantitative variable.

### Reproducing code example:

Here a rough initial implementation for example purposes. The final few lines are the only ones that differ from the current implementation.
```python
def max_distance_cumulative(data1, data2):
    """
    Computes the maximal vertical distance between cumulative distributions
    (this is the statistic for KS tests). Code mostly copied from
    scipy.stats.ks_twosamp

    Parameters
    ----------
    data1 : array_like
        First data set
    data2 : array_like
        Second data set
    Returns
    -------
    d : float
        Max distance, i.e. value of the Kolmogorov Smirnov test. Sign is + if
        the cumulative of data1 < the one of data2 at that location, else -.
    x : float
        Value of x where maximal distance d is reached.
    """
    from numpy import ma

    (data1, data2) = (ma.asarray(data1), ma.asarray(data2))
    (n1, n2) = (data1.count(), data2.count())
    mix = ma.concatenate((data1.compressed(), data2.compressed()))
    mixsort = mix.argsort(kind='mergesort')
    csum = np.where(mixsort < n1, 1./n1, -1./n2).cumsum()

    # Check for ties
    if len(np.unique(mix)) < (n1+n2):
        ind = np.r_[np.diff(mix[mixsort]).nonzero()[0], -1]
        csum = csum[ind]
        mixsort = mixsort[ind]

    csumabs = ma.abs(csum)
    i = csumabs.argmax()

    d = csum[i]
    # mixsort[i] contains the index of mix with the max distance
    x = mix[mixsort[i]]

    return (d, x)

```

I'm happy to make a PR if there is interest for this.

### PR DESCRIPTION
**Pull request by iosonofabio**
#### Reference issue
Aims to close #9389

#### What does this implement/fix?
Moves the result of Kolmogorov-Smirnov tests (1 or 2 samples) to a "tuple bunch" with two additional attributes:
- `statistic_location` indicating at what value of the stochastic variable (approximately) the maximal distance in CDF is reached
- `statistic_sign`, which is only nonempty for `alternative="two-sided"`, indicating whether the first distribution (for 2samp)/the empirical distribution (for 1samp) was above or below the second/theoretical distribution at the `statistic_location`. This is a cheap way to determine not only whether the distros are different, but also if they are, in what direction.

#### Additional information
Matt and I quickly discussed part 1, the `statistic_location`. I later realized having part 2 i.e. the sign is equally important for me and perhaps for other folks, and is a cheap addition in terms of performance.

I haven't adapted the part of the docstring that represents doctests yet, happy to do so but perhaps it'd be useful if one of the maintainers took a quick look the the draft PR first so I know if there are any major issues.

Thank you for your time and patience

### CODE REVIEW

**Comment by mdhaber** [comment]
@tupui @tirthasheshpatel @chrisb83 If you agree that this looks like a good idea, I can review this. There was a lot of user support in the original issue gh-9389, and they seem happy to have it added to the existing function, so I think it's fine to add these attributes rather than create a new function.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:8170]:
```suggestion
    return KstestResult(d, prob, statistic_location=d_location,
                        statistic_sign=d_sign)
```
PEP8: `scipy/stats/_stats_py.py:8170:80: E501 line too long (86 > 79 characters)`

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7552]:
```suggestion
KstestResult = _make_tuple_bunch('KstestResult', ['statistic', 'pvalue'],
```
I think this is causing most of the test failures, if not all.

**Comment by mdhaber** [comment]
OK, looks like there's just one real failure now:
https://github.com/scipy/scipy/blob/b5a805248c5c3b13e8d1b6237209854b6d0b4359/scipy/stats/tests/test_mstats_basic.py#L1877-L1882

This `Ks_2sampResult`:
https://github.com/scipy/scipy/blob/2e5883ef7af4f5ed4a5b80a1759a45e43163bf3f/scipy/stats/_stats_py.py#L7489

is no longer used anywhere, so I'd suggest just deleting the definition of it and deleting the test. That looks like it will fix the remaining failures.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7747]: What do you think about indicating the sign with `1` and `-1` (number) instead of a string? 
Also, even though it's redundant with the specified `alternative`, I'd suggest including it in the result for the one-sided alternative.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:8130]:
```suggestion
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:8113]:
```suggestion
    if alternative == 'less' or minS > maxS:
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:8117]:
```suggestion
    elif alternative == 'greater' or minS <= maxS:
```
Could just be `else`.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7638]: "closest to"? 
The KS statistics is the difference between the theoretical and empirical CDFs _at_ this location, no?

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7640]: I will probably have a suggested wording for this, too. I don't want to write a concrete suggestion yet, but I'm thinking we'll want to say something like `+1` if the statistic is the maximum _positive_ difference between the ____ CDF and the ____ CDF (can't remember which is theoretical vs empirical) and `-1` if the statistic is the maximum _negative_ difference...

**Comment by tupui** [comment]
I am +0.5. I personally would go with a separate function. We don't really have a lot of extra info in other tests. Instead we could have a general function to get the distance between 2 distributions (our API is slim there) and it would make perfect sense to have the location there.

File: scipy/stats/_stats_py.py
Comment (iosonofabio) [reply scipy/stats/_stats_py.py:7727]: Hehe let me take this out

File: scipy/stats/_stats_py.py
Comment (iosonofabio) [reply scipy/stats/_stats_py.py:7747]: Sure

File: scipy/stats/_stats_py.py
Comment (iosonofabio) [reply scipy/stats/_stats_py.py:7638]: It's mostly wording to ward off people who are worried about aliasing. Happy to rephrase more directly.

**Comment by mdhaber** [comment]
One part of my hesitation with adding a separate function is that if we were to add a function for the statistic used in each of our hypothesis tests, we'd double the number of functions in `stats`. I doubt we want to head in that direction, and I don't think the KS statistic is so special that is it deserves a function while the rest do not.

Also, it makes me think of Halton/Hammersley - those are actually different, yet we're talking about combining them. Here, we'd duplicate the function that computes the statistic, and the only difference between the functions would be that one version has the location of the maximum difference and the other has a p-value?

**Comment by tupui** [comment]
I agree that we don't want to duplicate the API with a location function for all tests. 

Here it's just that for this specific test we do something very generic and useful otherwise which is to compute the distance between CDFs.

I am fine doing both as well. Adding now an output for the location and maybe latter have a separate distance function.

**Comment by iosonofabio** [comment]
We could make an optional input argument called `compute_Pvalue=True` that skips the last part and is used just to compute the max distance and its location. Would that be a useful compromise in your opinion or just a mess? :smirk:

**Comment by mdhaber** [comment]
Mmm I'm more hesitant to add inputs to functions than outputs, and even more hesitant to add an input that instructs the function to add an output : )

It sounds like we have an agreement here - what @tupui said.
> Adding now an output for the location and maybe latter have a separate [CDF] distance function.

**Comment by mdhaber** [comment]
So to sum up @iosonofabio we can continue with this PR. Please consider the adjustments recommended above and I can take a second pass. I'd also suggest adding a unit test in `test_stats.py`. a good test would be to confirm that the statistic and p-value result object can still be unpacked and match the attributes, and also that the new attributes are present and appropriate (if possible).

**Comment by iosonofabio** [comment]
No problem, I'll get it done as soon as I get back to Oz (haven't slept in a while now...)

Thank you!

**Comment by mdhaber** [comment]
I saw you posted some new commits - did you want this to be reviewed again?

**Comment by iosonofabio** [comment]
Thanks @mdhaber - I was just overwhelmed in the last few days.

I have now added decent test functions, improved docstrings a little, and changed a piece of logic that you suggested and I believe was incorrect (please double check if you have time!).

If it passes CI, it is ready for you to take a look, thank you! Otherwise I'll just fix the errors first.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7646]:
```suggestion
            1 if the KS statistic is the maximal positive difference between
            empirical and hypothesized distributions (D+), -1 if the KS
            statistic is the negative difference (D-).
```
Because "D" is ultimately either D+ or D-, do you think we should alter the description above to eliminate "D"?

> KS test statistic, either D+ or D- (depending on the value of 'alternative')

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7972]:
```suggestion
            +1 if the empirical distribution function of `data1` exceeds
            the empircal distribution function of `data2` at 
            `statistic_location`, otherwise -1.
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7969]:
```suggestion
            Value from `data1` or `data2` corresponding with the KS statistic;
            i.e., the distance between the empirical distribution functions is
            greatest at this observation.
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7642]: Also, what do you think about "observation" instead of "location"? This is reinforce the idea that this is an element from the sample.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4013]: I'm not sure this is the strongest test. If you already thought through each one and reasoned out the answer (rather than just relying on the output of `ks_2samp` to tell you what the reference values should be), we can leave it.

Personally, though, I'd prefer to construct a test that makes it obvious what the answers should be. For instance, perhaps we would construct two samples that correspond very closely except at one observation. 
```python3
x = np.arange(10, dtype=np.float64)
y = x.copy()
y[6] = 6.9
stats.ks_2samp(x, y)
```

Something similar could be done for `ks_1samp`.

**Review by mdhaber — COMMENTED** [review]
Thanks @iosonofabio. The implementation looks perfect. My only comments are about the documentation and the tests.

File: scipy/stats/_stats_py.py
Comment (iosonofabio) [reply scipy/stats/_stats_py.py:8254]: yes

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:7643]: Is this what you wanted `statistic_location` for `ks_1samp` to be?

It's not that. Instead, it is the CDF at which the statistic is measured.

**Comment by mdhaber** [comment]
Please see iosonofabio/scipy#1.

File: scipy/stats/_stats_py.py
Comment (iosonofabio) [reply scipy/stats/_stats_py.py:7643]: Oops, that explains why the ks_2samp tests are fine but I couldn't get the one sample tests to work... Will fix tonight, thank you

File: scipy/stats/tests/test_stats.py
Comment (iosonofabio) [reply scipy/stats/tests/test_stats.py]: done now

**Comment by iosonofabio** [comment]
Hey @mdhaber - thank you again for your PR. I merged it and added similar tests for the one-sample KS test.

I think this is ready to go in if you guys approve. Also happy to make other changes or, if you prefer, feel free to push commits into this PR directly :smile:

**Comment by iosonofabio** [comment]
Sorry Matt, was taken in by some stuff! Let me check test results properly and I'll ask for review at that point. A day or two.

Thank you

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:8259]:
```suggestion
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:8265]: Please pick one.
If the latter, consider that the second sentence begins with "if a 2-sample test is run", but there is no corresponding part of the first sentence.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:3987]:
```suggestion
    @pytest.mark.parametrize("alternative, x56val, ref_location, ref_sign",
                             [('greater', 0.5, 0.5, +1),
                              ('less', 0.7, 0.7, -1),
                              ('two-sided', 0.5, 0.5, +1),
                              ('two-sided', 0.7, 0.7, -1)])
    def test_location_sign(self, alternative, x56val, ref_location, ref_sign):
        # Test that location and sign corresponding with statistic are as
        # expected. (Test is designed to be easy to predict.)
        x = np.arange(1.0, 11, dtype=np.float64) / 10.0
        x[5:7] = x56val
```

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:3991]:
```suggestion
        assert_array_almost_equal(res_list, expected_list)
```
This is not needed here or elsewhere.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4000]:
```suggestion
                 mode='auto'):
```
?

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:3899]: Is this needed?

**Comment by iosonofabio** [comment]
Thanks @mdhaber - implemented all requested changes. No further edits necessary AFAICT.

**Review by mdhaber — COMMENTED** [review]
It looks like the implementation in `ks_1samp` has been corrected and documentation is better. The main comments are about the tests. It seems that what needs to be tested is essentially the same for `kstest`, `ks_1samp`, and `ks_2samp`, but what is actually tested differs. Try to simplify the changes. We should see some symmetry in the testing (whetever is done for `ks_1samp` has a parallel for `ks_2ssamp`, and somehow we see that `kstest` is the same as those two).

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4176]: But why add this? It's shown that these attributes exist in both `test_location_sign` and `test_nametuples_agree`.

Comment (iosonofabio) [reply scipy/stats/tests/test_stats.py:4176]: You're right, of course. It's just that I have to watch my baby while coding, which makes these things hard to catch 😅

If you were as kind, it would be just fantastic if you could push this change to the PR branch and merge so we are fine with it. Otherwise, I'll try to find time next Wednesday to come back to this... Thank you!

**Comment by iosonofabio** [comment]
@mdhaber Thanks for the comment. I disagree that symmetry is essential. The 1-sample KS test is against a continuous distribution, whereas the 2-sample test is against a discretization thereof, which incurs all kinds of aliasing effects. I have done my best to try and satisfy the conflicting requirements of simplicity and theoretical clarity. As of now, I unfortunately have no time left for this PR. I'm not trying to be rude, but I really have no more time for this.

Given we are at the finish line, I would be grateful if you could push any changes that you think would improve the PR, symmetry or otherwise, and merge when you think it's good enough.

Thank you

**Comment by mdhaber** [comment]
The comment about some symmetry was resolved by the two recent commits. (It really crossed paths with the first of them, even though it shows up afterwards.)

File: scipy/stats/tests/test_mstats_basic.py
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:1877]: Rather than modifying this test, we can just remove it. It doesn't belong in `test_mstats_basic.py` if it's about `stats` objects anyway, and `Ks_2sampResult` is no longer used either in `stats` or `mstats`, so I think it's safe to remove this.

**Review by mdhaber — APPROVED** [review]

**Review by mdhaber — APPROVED** [review]
Approved, but I'd like a second opinion before merging, since I had a lot to do with the documentation and tests.

**Review by tirthasheshpatel — APPROVED** [review]
Looks good to me too! Thanks @iosonofabio @mdhaber! I initially thought that a new function would be better but, looking at this PR, it seems like either work just fine (with respect to API design choice). Since I haven't gone through all the reviews and comments on this PR, I would leave the decision to merge on @mdhaber.

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel! @tupui I'm going to merge now but if you have a chance for a quick glance, let me know if there is anything you'd like to change before the release and I'll open another PR.

**Comment by iosonofabio** [comment]
Thanks @mdhaber and everyone for your hard work and guidance!

**Pull request merged**
